pyvo/dal/adhoc.py

```diff
@@ -154,12 +168,68 @@ class DatalinkResultsMixin(AdhocServiceResultsMixin):
     """
     Mixing for datalink functionallity for results classes.
     """
+
     def iter_datalinks(self):
         """
         Iterates over all datalinks in a DALResult.
         """
-        for record in self:
-            yield record.getdatalink()
+        # To reduce the number of calls to the Datalink service, multiple
+        # IDs are sent in batches. The appropriate batch size is not available
+        # as each service has its own criteria for determining the maximum
+        # number of IDs processed per request. To overcome this limitation,
+        # this method initially sends all the available IDs and if the
+        # response is partial, uses the size of the response as the batch
+        # size.
+        if not hasattr(self, '_datalink'):
```

Comment:
As this is a generator, wouldn't it be better to turn those attributes into local variables?

Reply:
I don't see any harm in having `_datalink` (pointer to service) as a global for `DatalinkResult`. It is an attribute to that response.

Reply:
`_datalink` as a attribute of the result may be fine, but not `_current_batch`. When someone creates two generator instances for example.

Reply:
Doh! You are perfectly right! I've missed that somehow - but fixed now. Very good point.